### PR TITLE
Add Resample and AsType

### PR DESCRIPTION
**Pull request by rhoadesScholar**
Adds "Resample" node to resample arrays to a user-designated voxel size, using  skimage.transform.rescale

Pre-Merge Checklist:

- [ ] if this PR adds a feature: request merge into the latest development branch (`vX.Y-dev`)
- [ ] if this PR fixes a bug: request merge into the latest patch branch (`patch-X.Y.Z`)
- [ ] PR branch name is short and describes the feature/bug addressed in this PR
- [x] commit messages [are consistent](https://chris.beams.io/posts/git-commit/)
- [ ] changes reviewed by another contributor
- [x] tests cover changes
- [x] all tests pass

### CODE REVIEW

File: gunpowder/version_info.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 __major__   = 1
 __minor__   = 2
 __patch__   = 2
-__tag__     = ''
+__tag__     = '-raygun-dev'
```
Comment (funkey) [inline gunpowder/version_info.py:4]: That should probably not be part of the PR.

**Review by funkey — CHANGES_REQUESTED** [review]
1. Could you please write a test for `Resample` and `AsType`?
2. Please change the name of the PR to "Add Resample and AsType" or similar

File: gunpowder/version_info.py (hunk 1 above)
Comment (rhoadesScholar) [reply gunpowder/version_info.py:4]: Agreed. I had it in during dev, and forgot to remove it. Sorry about that!

**Comment by rhoadesScholar** [comment]
Tests written for Resample and AsType.

File: gunpowder/torch/nodes/train.py [hunk 2]
```diff
@@ -315,6 +315,20 @@ def train_step(self, batch, request):
 
         if self.summary_writer and batch.iteration % self.log_every == 0:
             self.summary_writer.add_scalar("loss", batch.loss, batch.iteration)
+            if hasattr(self.loss, 'loss_dict'):
```
Comment (pattonw) [inline gunpowder/torch/nodes/train.py:318]: This section doesn't seem to be relevant to adding Resample/AsType

File: tests/cases/astype.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    def provide(self, request):
+
+        batch = Batch()
+
+        # have the pixels encode their position
+        for (array_key, spec) in request.array_specs.items():
+
+            roi = spec.roi
+
+            for d in range(3):
```
Comment (pattonw) [inline tests/cases/astype.py:30]: I think these asserts are redundant. Gunpowder should make sure that any request that is made is valid according to the `ArraySpec` for any `ArrayKey` that is provided by this node.

File: tests/cases/astype.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                        range(roi.get_begin()[2], roi.get_end()[2]), indexing='ij')
+                data = meshgrids[0] + meshgrids[1] + meshgrids[2]
+
+                self.assertTrue(np.array_equal(array.data, data), str(array_key))
+
+            elif array_key == ArrayKeys.RAW_TYPECAST:
+
+                self.assertTrue(array.data.dtype == np.float16, 
+                                f'RAW_TYPECAST dtype: {array.data.dtype} does not equal expected: np.float16')
+                self.assertTrue(int(array.data[1,11,1]) == 43,
```
Comment (pattonw) [inline tests/cases/astype.py:91]: 43 or 13?

File: tests/cases/resample.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                        range(roi.get_begin()[0], roi.get_end()[0]),
+                        range(roi.get_begin()[1], roi.get_end()[1]),
+                        range(roi.get_begin()[2], roi.get_end()[2]), indexing='ij')
+                data = meshgrids[0] + meshgrids[1] + meshgrids[2]
+
+                self.assertTrue(np.array_equal(array.data, data), str(array_key))
+
+            elif array_key == ArrayKeys.RAW_RESAMPLED:
+
+                self.assertTrue(array.data[0,0,0] == 31,
```
Comment (pattonw) [inline tests/cases/resample.py:89]: Not sure if I follow why the expected value is 31 instead of 30. Since you request array of shape (120,120,120), and a total request size of (200,200,200) I would expect this array to be offset by 40 on each axis. voxel-size of 4 means an offset of 10 on each axis. Adding those up I would expect a value of 30.

File: gunpowder/nodes/__init__.py [hunk 6]
```diff
@@ -28,7 +29,9 @@
 from .random_provider import RandomProvider
 from .rasterize_graph import RasterizationSettings, RasterizeGraph
 from .reject import Reject
+from .reject_empty import RejectEmpty
```
Comment (pattonw) [inline gunpowder/nodes/__init__.py:32]: Add RejectEmpty not in pull request title

File: gunpowder/nodes/astype.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+from .batch_filter import BatchFilter
+from gunpowder.array import ArrayKey, Array
+from gunpowder.batch import Batch
+import logging
+
+logger = logging.getLogger(__name__)
+
+class AsType(BatchFilter):
```
Comment (pattonw) [inline gunpowder/nodes/astype.py:8]: This could probably be done dynamically. Since the batch request can contain metadata, instead of providing the `target_dtype` on initialization of the `AsType` node, one option would be to look at the request metadata and change the dtype to whatever has been requested.
Not sure if there would be a use for that. Might over complicate the problem.

Comment (pattonw) [reply gunpowder/nodes/astype.py:8]: Maybe @funkey has some thoughts.
Otherwise `AsType` as is looks good

**Comment by rhoadesScholar** [comment]
Pull Request out of date and the main author on paternity leave until 01/17/2022. Will be splitting off updates unrelated to AsType and Resample nodes at that time into a separate pull request.

Thank you for your feedback/interest.

File: gunpowder/nodes/astype.py (hunk 7 above)
Comment (funkey) [reply gunpowder/nodes/astype.py:8]: I'd leave it as is. Automatically changing the `dtype` based on the request is a bit too much magic.

File: tests/cases/resample.py (hunk 5 above)
Comment (rhoadesScholar) [reply tests/cases/resample.py:89]: Yup. ["Behavior is not a pixel-perfect copy of down/upsample nodes, because this node relies on skimage.transform.rescale to perform non-integer scaling factors."](https://github.com/funkey/gunpowder/blob/445d2ed875e2d33de28ff786c9a9767ea84049f9/gunpowder/nodes/resample.py#L14) I believe the skimage function differs in how it localizes data, noting "that 0th pixel is at position (0.5, ...)" 
See [the source for reference.](https://github.com/scikit-image/scikit-image/blob/55dee6421c1411388eb299f70244c90d956140ef/skimage/transform/_warps.py#L224)

Please let me know if you think this is a significant issue that needs addressing, and if you have any suggestions for elegant solutions. Thank you!

File: tests/cases/astype.py (hunk 4 above)
Comment (rhoadesScholar) [reply tests/cases/astype.py:91]: Yikes! Typo! Good catch. Thanks!

File: gunpowder/torch/nodes/train.py (hunk 2 above)
Comment (rhoadesScholar) [reply gunpowder/torch/nodes/train.py:318]: Correct. I will remove some commits that were accidentally pushed to this PR. Thanks!

File: tests/cases/astype.py (hunk 3 above)
Comment (rhoadesScholar) [reply tests/cases/astype.py:30]: Potentially. This was included following [the example present in the pre-existing Downsample test](https://github.com/funkey/gunpowder/blob/445d2ed875e2d33de28ff786c9a9767ea84049f9/tests/cases/downsample.py#L30).

It seems like there is no harm in checking, but I will be happy to remove it if you think it muddies it. Thanks!

File: gunpowder/nodes/__init__.py (hunk 6 above)
Comment (rhoadesScholar) [reply gunpowder/nodes/__init__.py:32]: Correct. I will remove some commits that were accidentally pushed to this PR. Thanks!

Comment (rhoadesScholar) [reply gunpowder/nodes/__init__.py]: Fixed with commits: 
[e2d09e6](https://github.com/funkey/gunpowder/pull/163/commits/e2d09e6ee19f2481729c1b625c1b0b3cb472e97c)
[dac8d27](https://github.com/funkey/gunpowder/pull/163/commits/dac8d27ed2adea09637bc8819fc1364e9861f0ce)

File: gunpowder/torch/nodes/train.py (hunk 2 above)
Comment (rhoadesScholar) [reply gunpowder/torch/nodes/train.py]: Fixed with commits: 
[e2d09e6](https://github.com/funkey/gunpowder/pull/163/commits/e2d09e6ee19f2481729c1b625c1b0b3cb472e97c)
[dac8d27](https://github.com/funkey/gunpowder/pull/163/commits/dac8d27ed2adea09637bc8819fc1364e9861f0ce)

File: tests/cases/astype.py (hunk 4 above)
Comment (rhoadesScholar) [reply tests/cases/astype.py:91]: Fixed in commit [f4e9a058](https://github.com/funkey/gunpowder/pull/163/commits/f4e9a058a4a42bfdb726a088d5c73278c6ad63d2)

File: tests/cases/resample.py (hunk 5 above)
Comment (pattonw) [reply tests/cases/resample.py:89]: Oh nvm, it makes sense to me know. its averaging over the voxels in the raw roi: (40:48, 40:48, 40:48). Those values are (30,31,31,32,31,32,32,33). Average to 31.5, and casting to an int is rounding down for some reason. Wierd but ok. Maybe comment how you get to 31 for people who might come across this in the future and be confused like me.

File: tests/cases/astype.py (hunk 3 above)
Comment (pattonw) [reply tests/cases/astype.py:30]: Definitely not a big deal. I think we're good to merge with or without.
I'd vote for deleting it. Especially since it is already redundant in that test as well.

**Comment by pattonw** [comment]
Just a few quick changes and we can go ahead and merge:
1) Since you're adding new features can you target it at the patch-1.2.3? That will make it easier for @funkey when he eventually needs to update a new release.
2) Can you add these nodes to the docs. You can do that in the `docs/build/api.rst` file: https://github.com/funkey/gunpowder/blob/master/docs/build/api.rst

File: tests/cases/resample.py (hunk 5 above)
Comment (rhoadesScholar) [reply tests/cases/resample.py:90]: Commented in commit f2310f2.

**Comment by rhoadesScholar** [comment]
> Just a few quick changes and we can go ahead and merge:
> 
> 1. Since you're adding new features can you target it at the patch-1.2.3? That will make it easier for @funkey when he eventually needs to update a new release.
> 2. Can you add these nodes to the docs. You can do that in the `docs/build/api.rst` file: https://github.com/funkey/gunpowder/blob/master/docs/build/api.rst

1. Done.
2. Added in c04f5d7, but did not re-generate html.

**Comment by pattonw** [comment]
No need to re-generate the HTML, Jan will do that when he creates a new release.
This all looks good and ready to merge
